x/programs/cmd/simulator/src/lib.rs

```diff
@@ -48,6 +46,15 @@ pub struct Step {
     pub require: Option<Require>,
 }
 
+#[derive(Debug, Serialize, PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub struct SimulatorStep<'a> {
+    /// The key of the caller used in each step of the plan.
+    pub caller_key: &'a String,
```

Comment:
`String` is a smart-pointer to a heap allocation that effectively looks like this
```rust
struct String {
    ptr: *const u8,
    len: usize,
    capacity: usize,
}
```
(It doesn't actually look like that, but you can think of it as such)
`&str` is effectively a pointer and a length (`size_of::<&str>()` is 1/3 smaller).
`&String` is a pointer to a struct on the stack that points to a heap allocation; you have an extra indirection. So it's always better to use `&str` and `&[T]` instead of `&String` and `&Vec<T>`
```suggestion
    pub caller_key: &'a str,
```

Reply:
I see, so `&String` is most of the time an anti-pattern ! Thank you very much!